determined-ai/determined
# feat: checkpoint download through master

## Description



Users do not always have access to checkpoint storage, and hence downloading checkpoints with the CLI does not always work. This feature overcomes the limitation by supporting checkpoint downloads through the master since t is often easier to grant access to the master than to individual users.  This feature only supports AWS S3 for now and relies on implicit IAM permissions on the master.  In the future, it can be extended to support other checkpoint storage types and take explicitly provided credentials.

A single endpoints is added:
- `/checkpoints/:checkpoint_uuid/`

The implementation downloads the checkpoint by making API calls to the S3 endpoint, packages it as a tgz or a zip file, and then sends the file to the client in the HTTP response. The whole process operates as streaming. No temporary files are stored or cached.


*Note:* One of the integration tests needs to run in the upstream repo to have access to cloud resources. Hence this PR is also pushed to https://github.com/determined-ai/determined/tree/hanyu/master-checkpoint-download


## Test Plan



- Use `devcluster` to start a dev cluster. Configure the `checkpoint_storage` section to use an S3 bucket and provide appropriate AWS credentials. For example:
```yaml
checkpoint_storage:
  type: s3
  bucket: <your test S3 bucket>
  prefix: /devcluster/checkpoint-download-test
  access_key: *** (redacted) ***
  secret_key: *** (redacted) ***
```
- Make sure the same or equivalent credentials are available in environment variables or `~/.aws/credentials`.
- Run `export DET_MASTER=http://localhost:8080`
- Run a simple experiment, e.g., the MNIST example. When the experiment completes, a checkpoint should have been created.
- Obtain the checkpoint UUID under that experiment using the CLI.
- Go to the S3 bucket in the AWS Console to verify the checkpoint with the same UUID.
- Obtain the bearer token for the cluster.
- Use cURL to call the tgz endpoint. For example:
  Substitute `$TOKEN` and `$UUID`, and run:  `curl -v -o out.det-curl.tgz --header "Authorization: Bearer $TOKEN" --header "Accept: application/gzip" http://localhost:8080/checkpoints/$UUID`
- Repeat the last step for the tgz endpoint. For example:
  Substitute `$TOKEN` and `$UUID`, and run:  `curl -v -o out.det-curl.tgz --header "Authorization: Bearer $TOKEN" --header "Accept: application/zip" http://localhost:8080/checkpoints/$UUID`
- Extract both files -- make sure you do not overwrite them. 
- Verify the contents extracted from the tgz and zip files are the same as those in S3.


## Commentary (optional)




## Checklist
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.





[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
